### DISCUSSION TITLE
S3 bucket only keep specific  numbers of objects

### DISCUSSION THREAD

**Discussion by itsyosefali** [aws/aws-cdk#24224] [Q&A]
as you see from the title i want to only keep 3 newest objects of the bucket and delete the rest (the objects are backups that upload from my site automatically everyday and they dont have the same name ) 

i heard that i have to use lambda function but as i know i have to pay them per request 
so is there any help that i can do it with the policy or the life cycle  ?

**Comment by kaizencc** [comment]
Maybe you can play around with S3 Bucket Versioning: https://docs.aws.amazon.com/AmazonS3/latest/userguide/Versioning.html. The use case for object versioning seem similar to the use case you describe.

I'm sure you've looked into S3 Lifecycle Rules as well: https://docs.aws.amazon.com/AmazonS3/latest/userguide/intro-lifecycle-rules.html. I think you can play around with tags and SDK calls only, and specify a Lifecycle filter that's something like "if tagged with `delete`, then delete after X days". But then you'll have to run a script on a regular cadence to replace what a lambda function would do. 

AFAIK, the canonical way is to use lambdas here, but I think you can replace the lambda function with a script that hits some SDK calls if you really want to, and run that script on a different engine if you so choose.

**Reply by itsyosefali** [reply]
I've tried the versioning  but it didn't work because my bucket everyday i sent object to it that doesn't have the same name as the previous one so it's ussless at this point to use it

And yes it's really weird to not see this kind of thing like only keeping X number of objects in the bucket at the lifecycle rules

i don't know what is the best solution so far , im trying so hard to avoid the lambda function  but i guess there's no other way
